File: MapsSDKExamplesObjc/Search/SearchAdditionalDataViewController.m
/**
 * Copyright (c) 2019 TomTom N.V. All rights reserved.
 *
 * This software is the proprietary copyright of TomTom N.V. and its
 * subsidiaries and may be used for internal evaluation purposes or commercial
 * use strictly subject to separate licensee agreement between you and TomTom.
 * If you are the licensee, you are only permitted to use this Software in
 * accordance with the terms of your license agreement. If you are not the
 * licensee then you are not authorised to use this software in any manner and
 * should immediately return it to TomTom N.V.
 */

#import "SearchAdditionalDataViewController.h"
#import <TomTomOnlineSDKSearch/TomTomOnlineSDKSearch.h>

@interface PolygonAdditionalDataVisitior : NSObject <TTGeoJSONObjectVisitor, TTGeoJSONGeoVisitor>
@property(nonatomic, strong) NSMutableArray<TTGeoJSONLineString *> *lineStrings;
@end

@implementation PolygonAdditionalDataVisitior

- (instancetype)init {
    self = [super init];
    if (self) {
        self.lineStrings = [NSMutableArray new];
    }
    return self;
}

- (void)visitFeatureCollection:(TTGeoJSONFeatureCollection *)featureCollection {
    for (TTGeoJSONFeature *feature in featureCollection.features) {
        [feature visitResult:self];
    }
}

- (void)visitPolygon:(TTGeoJSONPolygon *)polygon {
    [self.lineStrings addObject:polygon.exteriorRing];
}

- (void)visitMultiPolygon:(TTGeoJSONMultiPolygon *)multiPolygon {
    for (TTGeoJSONPolygon *polygon in multiPolygon.polygons) {
        [self.lineStrings addObject:polygon.exteriorRing];
    }
}

@end

@interface SearchAdditionalDataViewController () <TTSearchDelegate, TTAdditionalDataSearchDelegate>
@property(nonatomic, strong) TTSearch *search;
@property(nonatomic, strong) TTAdditionalDataSearch *searchAdditionalData;
@end

@implementation SearchAdditionalDataViewController

- (OptionsView *)getOptionsView {
    return [[OptionsViewSingleSelect alloc] initWithLabels:@[ @"Poland", @"Amsterdam", @"Schiphol" ] selectedID:-1];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    self.search = [[TTSearch alloc] initWithKey:Key.Search];
    self.search.delegate = self;
    self.searchAdditionalData = [[TTAdditionalDataSearch alloc] initWithKey:Key.Search];
    self.searchAdditionalData.delegate = self;
}

#pragma mark OptionsViewDelegate

- (void)displayExampleWithID:(NSInteger)ID on:(BOOL)on {
    [super displayExampleWithID:ID on:on];
    [self.progress show];
    [self.mapView.annotationManager removeAllOverlays];
    switch (ID) {
    case 2:
        [self displayAdditionalDataLanduse];
        break;
    case 1:
        [self displayAdditionalDataAmsterdam];
        break;
    default:
        [self displayAdditionalDataPoland];
        break;
    }
}

#pragma mark Examples

- (void)displayAdditionalDataAmsterdam {
    TTSearchQuery *query = [[[TTSearchQueryBuilder createWithTerm:@"Amsterdam"] withIdxSet:TTSearchIndexGeographies] build];
    [self.search searchWithQuery:query];
}

- (void)displayAdditionalDataLanduse {
    TTSearchQuery *query = [[[TTSearchQueryBuilder createWithTerm:@"Amsterdam Airport Schiphol"] withIdxSet:TTSearchIndexPointOfInterest] build];
    [self.search searchWithQuery:query];
}

- (void)displayAdditionalDataPoland {
    TTSearchQuery *query = [[[TTSearchQueryBuilder createWithTerm:@"Poland"] withIdxSet:TTSearchIndexGeographies] build];
    [self.search searchWithQuery:query];
}

#pragma mark TTSearchDelegate
- (void)search:(TTSearch *)search completedWithResponse:(TTSearchResponse *)response {
    TTSearchResult *result = response.results.firstObject;
    if (!result) {
        return;
    }
    TTAdditionalDataSources *additionalData = result.additionalDataSources;
    if (!additionalData) {
        return;
    }
    TTGeometryDataSource *geometryData = additionalData.geometryDataSource;
    if (!geometryData) {
        return;
    }
    TTAdditionalDataSearchQuery *query = [[TTAdditionalDataSearchQueryBuilder createWithDataSource:geometryData] build];
    [self.searchAdditionalData additionalDataSearchWithQuery:query];
}

- (void)search:(TTSearch *)search failedWithError:(TTResponseError *)error {
    [self handleError:error];
}

#pragma mark TTAdditionalDataSearchDelegate
- (void)additionalDataSearch:(TTAdditionalDataSearch *)additionalDataSearch completedWithResponse:(TTAdditionalDataSearchResponse *)response {
    [self.progress hide];
    TTAdditionalDataSearchResult *result = response.results.firstObject;
    if (!result) {
        return;
    }
    PolygonAdditionalDataVisitior *visitor = [PolygonAdditionalDataVisitior new];
    [result visitGeoJSONResult:visitor];
    NSMutableArray<TTPolygon *> *mapPolygons = [NSMutableArray new];
    for (TTGeoJSONLineString *lineString in visitor.lineStrings) {
        TTPolygon *mapPolygon = [TTPolygon polygonWithCoordinatesData:lineString opacity:0.7 color:[TTColor Red] colorOutline:[TTColor Red]];
        [mapPolygons addObject:mapPolygon];
        [self.mapView.annotationManager addOverlay:mapPolygon];
    }
    [self.mapView zoomToCoordinatesDataCollection:mapPolygons];
}

- (void)additionalDataSearch:(TTAdditionalDataSearch *)additionalDataSearch failedWithError:(TTResponseError *)error {
    [self handleError:error];
}

@end

File: MapsSDKExamplesSwift/Search/SearchAdditionalDataViewController.swift
/**
 * Copyright (c) 2019 TomTom N.V. All rights reserved.
 *
 * This software is the proprietary copyright of TomTom N.V. and its subsidiaries and may be used
 * for internal evaluation purposes or commercial use strictly subject to separate licensee
 * agreement between you and TomTom. If you are the licensee, you are only permitted to use
 * this Software in accordance with the terms of your license agreement. If you are not the
 * licensee then you are not authorised to use this software in any manner and should
 * immediately return it to TomTom N.V.
 */

import MapsSDKExamplesCommon
import MapsSDKExamplesVC
import TomTomOnlineSDKMaps
import TomTomOnlineSDKSearch
import UIKit

class SearchAdditionalDataViewController: MapBaseViewController, TTSearchDelegate, TTAdditionalDataSearchDelegate {
    let search = TTSearch()
    let searchAdditionalData = TTAdditionalDataSearch()

    override func getOptionsView() -> OptionsView {
        return OptionsViewSingleSelect(labels: ["Poland", "Amsterdam", "Schiphol"], selectedID: -1)
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        search.delegate = self
        searchAdditionalData.delegate = self
    }

    // MARK: OptionsViewDelegate

    override func displayExample(withID ID: Int, on: Bool) {
        super.displayExample(withID: ID, on: on)
        progress.show()
        mapView.annotationManager.removeAllOverlays()
        switch ID {
        case 2:
            displayAdditionalDataLanduse()
        case 1:
            displayAdditionalDataAmsterdam()
        default:
            displayAdditionalDataPoland()
        }
    }

    // MARK: Examples

    func displayAdditionalDataAmsterdam() {
        let query = TTSearchQueryBuilder.create(withTerm: "Amsterdam")
            .withIdxSet(.geographies)
            .build()
        search.search(with: query)
    }

    func displayAdditionalDataLanduse() {
        let query = TTSearchQueryBuilder.create(withTerm: "Amsterdam Airport Schiphol")
            .withIdxSet(.pointOfInterest)
            .build()
        search.search(with: query)
    }

    func displayAdditionalDataPoland() {
        let query = TTSearchQueryBuilder.create(withTerm: "Poland")
            .withIdxSet(.geographies)
            .build()
        search.search(with: query)
    }

    // MARK: TTSearchDelegate

    func search(_: TTSearch, completedWith response: TTSearchResponse) {
        guard let result = response.results.first else {
            return
        }
        guard let additionalData = result.additionalDataSources else {
            return
        }
        guard let geometryData = additionalData.geometryDataSource else {
            return
        }
        let query = TTAdditionalDataSearchQueryBuilder.create(with: geometryData).build()
        searchAdditionalData.additionalDataSearch(with: query)
    }


    func search(_: TTSearch, failedWithError error: TTResponseError) {
        handleError(error)
    }

    // MARK: TTAdditionalDataSearchDelegate

    func additionalDataSearch(_: TTAdditionalDataSearch, completedWith response: TTAdditionalDataSearchResponse) {
        progress.hide()
        guard let result = response.results.first else {
            return
        }
        let visitor = PolygonAdditionalDataVisitior()
        result.visitGeoJSONResult(visitor)
        var mapPolygons: [TTPolygon] = []
        for lineString in visitor.lineStrings {
            let mapPolygon = TTPolygon(coordinatesData: lineString, opacity: 0.7, color: TTColor.Red(), colorOutline: TTColor.Red())
            mapPolygons.append(mapPolygon)
            mapView.annotationManager.add(mapPolygon)
        }
        mapView.zoom(toCoordinatesDataCollection: mapPolygons)
    }


    func additionalDataSearch(_: TTAdditionalDataSearch, failedWithError error: TTResponseError) {
        handleError(error)
    }
}

class PolygonAdditionalDataVisitior: NSObject, TTGeoJSONObjectVisitor, TTGeoJSONGeoVisitor {
    var lineStrings: [TTGeoJSONLineString] = []

    func visit(_ featureCollection: TTGeoJSONFeatureCollection) {
        for feature in featureCollection.features {
            feature.visitResult(self)
        }
    }

    func visit(_ polygon: TTGeoJSONPolygon) {
        lineStrings.append(polygon.exteriorRing)
    }

    func visit(_ multiPolygon: TTGeoJSONMultiPolygon) {
        for polygon in multiPolygon.polygons {
            lineStrings.append(polygon.exteriorRing)
        }
    }
}
